src/DataWrangling/ecco_metadata.jl

```diff
@@ -0,0 +1,173 @@
+# - `dates`: The dates of the dataset, in a `AbstractCFDateTime` format.
+# - `version`: The version of the dataset, could be ECCO2Monthly, ECCO2Daily, or ECCO4Monthly.
+struct ECCOMetadata{D, V} 
+    name  :: Symbol
+    dates :: D
+    version :: V
+end
+
+Base.show(io::IO, metadata::ECCOMetadata) = 
+    print(io, "ECCOMetadata:", '\n',
+    "├── field: $(metadata.name)", '\n',
+    "├── dates: $(metadata.dates)", '\n',
+    "└── data version: $(metadata.version)")
+
+# The default is the ECCO2Daily dataset at 1993-01-01.
+function ECCOMetadata(name::Symbol; 
+                      date = DateTimeProlepticGregorian(1993, 1, 1),
+                   version = ECCO2Daily()) 
+             
+    return ECCOMetadata(name, date, version) 
+end
+
+# Treat ECCOMetadata as an array to allow iteration over the dates.
+Base.getindex(metadata::ECCOMetadata, i::Int) = @inbounds ECCOMetadata(metadata.name, metadata.dates[i], metadata.version)
+Base.length(metadata::ECCOMetadata)       = length(metadata.dates)
+Base.eltype(metadata::ECCOMetadata)       = Base.eltype(metadata.dates)
+Base.first(metadata::ECCOMetadata)        = @inbounds ECCOMetadata(metadata.name, metadata.dates[1], metadata.version)
+Base.last(metadata::ECCOMetadata)         = @inbounds ECCOMetadata(metadata.name, metadata.dates[end], metadata.version)
+Base.iterate(metadata::ECCOMetadata, i=1) = (@inline; (i % UInt) - 1 < length(metadata) ? (@inbounds ECCOMetadata(metadata.name, metadata.dates[i], metadata.version), i + 1) : nothing)
+
+Base.axes(metadata::ECCOMetadata{<:AbstractCFDateTime})    = 1
+Base.first(metadata::ECCOMetadata{<:AbstractCFDateTime})   = metadata
+Base.last(metadata::ECCOMetadata{<:AbstractCFDateTime})    = metadata
+Base.iterate(metadata::ECCOMetadata{<:AbstractCFDateTime}) = (metadata, nothing)
+Base.iterate(::ECCOMetadata{<:AbstractCFDateTime}, ::Any)  = nothing
+
+Base.size(data::ECCOMetadata{<:Any, <:ECCO2Daily})   = (1440, 720, 50, length(data.dates))
+Base.size(data::ECCOMetadata{<:Any, <:ECCO2Monthly}) = (1440, 720, 50, length(data.dates))
+Base.size(data::ECCOMetadata{<:Any, <:ECCO4Monthly}) = (720,  360, 50, length(data.dates))
+
+Base.size(::ECCOMetadata{<:AbstractCFDateTime, <:ECCO2Daily})   = (1440, 720, 50, 1)
+Base.size(::ECCOMetadata{<:AbstractCFDateTime, <:ECCO2Monthly}) = (1440, 720, 50, 1)
+Base.size(::ECCOMetadata{<:AbstractCFDateTime, <:ECCO4Monthly}) = (720,  360, 50, 1)
+
+# The whole range of dates in the different dataset versions
+all_ecco_dates(::ECCO4Monthly) = DateTimeProlepticGregorian(1992, 1, 1) : Month(1) : DateTimeProlepticGregorian(2023, 12, 1)
+all_ecco_dates(::ECCO2Monthly) = DateTimeProlepticGregorian(1992, 1, 1) : Month(1) : DateTimeProlepticGregorian(2023, 12, 1)
+all_ecco_dates(::ECCO2Daily)   = DateTimeProlepticGregorian(1992, 1, 4) : Day(1)   : DateTimeProlepticGregorian(2023, 12, 31)
+
+# File name generation specific to each Dataset version
+function file_name(metadata::ECCOMetadata{<:AbstractCFDateTime, <:ECCO4Monthly})
```

Comment:
we use `filename` almost exclusively everywhere so I suggest keeping that style here. (technically `filename` is one word, but mainly I think we just want to be consistent)

Reply:
not sure, `filename` is a variable given as a keyword argument and `file_name` is a function with `metadata` as an input. We can change the name of the function but `filename` should be the variable in my opinion. In JRA55 the pattern is `filename = filenames[variable]` (a dictionary) but I don't think that is a pattern to take example from because it does not generalize to multiple files. 

I wanted to change JRA55 to allow multiple year forcing, in that case we have to come up with a similar pattern of metadata and `function how_we_want_to_call_it(metadata) = the_filename` but we should do that in a new PR.

Reply:
What about `filename = metadata_filename(metadata)`

Reply:
`metadata_filename` is good.